cpmpy/solvers/scip.py

```diff
@@ -0,0 +1,419 @@
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            Examples of scip supported arguments include:
+                - TODO
+
+            For a full list of scip parameters, please visit 
+            https://www.scipopt.org/doc/html/PARAMETERS.php
+        """
+        #import pyscipopt as scip
+
+        if time_limit is not None:
+            # in seconds
+            self.scip_model.setParam("limits/time", time_limit)
+
+        if solution_callback is not None:
+            raise Exception("SCIP: solution callback not (yet?) implemented")
+
+        # call the solver, with parameters
+        #for param, val in kwargs.items():
+        #    self.scip_model.setParam(param, val)
+        self.scip_model.setParams(kwargs)
+
+        _ = self.scip_model.optimize()
+        scip_objective = self.scip_model.getObjective()
+
+        scip_status = self.scip_model.getStatus()
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.scip_model.getSolvingTime()
+
+        # translate exit status
+        if scip_status == "optimal":
+            if self.has_objective():
+                self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+            else:
+                self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif scip_status == "infeasible":
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif scip_status.endswith("limit"):  # timelimit, nodelimit, ...
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif scip_status == "unknown":
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN
+        # TODO: "userinterrupt" and check if there is a solution?
+        else: # there are many more, e.g. one for each limit
+            raise NotImplementedError(
+                f"Translation of scip status {scip_status} to CPMpy status not implemented")  # a non-mapped status type, please report on github
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return(self.cpm_status)
+
+        # translate solution values (of user specified variables only)
+        self.objective_value_ = None
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                solver_val = self.scip_model.getVal(self.solver_var(cpm_var))
+                if cpm_var.is_bool(): 
+                    cpm_var._value = solver_val >= 0.5
+                else:
+                    cpm_var._value = int(solver_val)
+            # set _objective_value
+            if self.has_objective():
+                self.objective_value_ = scip_objective.getObjVal()
+
+        self.scip_model.freeTransform()  # Mark Turner from SCIP told me you need to do this if you want to support adding additional vars/constraints later...
```

Comment:
So a tiny summary for this: SCIP transforms the problem that you are actual solving into something it believes is easier to solve (presolve techniques etc). All solutions when found are then transformed back to the original space, because that is obviously what the user modelled and can interpret. Allowing the user to change the model during solving gets messy when balancing these things however, so it's forbidden. What ```self.scip_model.freeTransform()``` does is to remove all information from the transformed space, and just leave the original model. All solutions are kept, and the user can now change the model as they will. The downside is that potentially useful information for speeding up the next optimisation call is thrown out.